### PR TITLE
Support identifying the underlying file and path of a given element.

### PR DESCRIPTION
**Pull request by seanparsons**
**Problem:**
We want to be able to edit a component that is shown in the canvas for a given file but is not from that file, but have no way to "find" what element it is in which particular file.

**Fix:**
Provide functionality for identifying the originating file for a given element along with the component it relates to in that file.

**Commit Details:**
- Copied some path related functions and added `absolutePathFromRelativePath`.
- Added `findElementWithUID` to lookup a given element given the `data-uid`
  that relates to it.
- Added `dropFirstScenePathElement` to assist with walking a path down
  the various scenes that are included.
- Added `importedFromWhere` to find the import that a variable name relates to.
- Implemented `normalisePathToUnderlyingTarget` which performs the work
  of jumping around the project contents of a project finding the underlying
  components as it goes.

### CODE REVIEW

File: editor/src/components/custom-code/code-file.ts [hunk 1]
```diff
@@ -254,3 +273,169 @@ export function codeCacheToBuildResult(cache: {
...
+}
+
+export function normalisePathError(errorMessage: string): NormalisePathError {
+  return {
+    type: 'NORMALISE_PATH_ERROR',
+    errorMessage: errorMessage,
+  }
+}
+
+export interface NormalisePathUnableToProceed {
```
Comment (Rheeseyb) [inline editor/src/components/custom-code/code-file.ts:321]: Seems like this could do with coming with an error message too?

File: editor/src/components/custom-code/code-file.ts [hunk 2]
```diff
@@ -254,3 +273,169 @@ export function codeCacheToBuildResult(cache: {
...
+      const staticPath = TP.dynamicPathToStaticPath(elementPath)
+      const potentiallyDroppedFirstSceneElementResult = TP.dropFirstScenePathElement(staticPath)
+      if (potentiallyDroppedFirstSceneElementResult.droppedScenePathElements == null) {
+        // As the scene path is empty, there's no more traversing to do, the target is in this file.
+        return normalisePathSuccess(staticPath, currentFilePath, currentFile)
+      } else {
+        const droppedPathPart = potentiallyDroppedFirstSceneElementResult.droppedScenePathElements
+        if (droppedPathPart.length === 0) {
+          return normalisePathError(
+            `Unable to handle empty scene path part for ${TP.toString(elementPath)}`,
```
Comment (Rheeseyb) [inline editor/src/components/custom-code/code-file.ts:372]: Is this the difference between `scenePath([])` and `scenePath([[]])`?

File: editor/src/components/custom-code/code-file.ts [hunk 3]
```diff
@@ -254,3 +273,169 @@ export function codeCacheToBuildResult(cache: {
...
+            }
+            // Handle things like divs.
+            if (isIntrinsicElement(targetElement.name)) {
+              return normalisePathSuccess(
+                TP.dynamicPathToStaticPath(potentiallyDroppedFirstSceneElementResult.newPath),
+                currentFilePath,
+                currentFile,
+              )
+            } else {
+              return lookupElementImport(targetElement.name.baseVariable)
```
Comment (Rheeseyb) [inline editor/src/components/custom-code/code-file.ts:432]: This needs to be extended to support the fully qualified element name (I think this would break right now on `animated.div`)

Comment (Rheeseyb) [reply editor/src/components/custom-code/code-file.ts:432]: Oh sorry, I'm an idiot. I just realised what `baseVariable` is :facepalm:

File: editor/src/core/shared/template-path.ts [hunk 4]
```diff
@@ -914,3 +914,35 @@ export function staticScenePathContainsElementPath(
...
   })
 }
+
+export function isScenePathEmpty(path: TemplatePath): boolean {
+  if (isScenePath(path)) {
+    return path.sceneElementPaths.length === 0
+  } else {
+    return path.scene.sceneElementPaths.length === 0
+  }
+}
```
Comment (Rheeseyb) [inline editor/src/core/shared/template-path.ts:924]: Nice!

File: editor/src/core/shared/template-path.ts [hunk 5]
```diff
@@ -914,3 +914,35 @@ export function staticScenePathContainsElementPath(
...
+
+interface DropFirstScenePathElementResultType {
+  newPath: InstancePath
+  droppedScenePathElements: StaticElementPath | null
+}
+
+export function dropFirstScenePathElement(
+  path: StaticInstancePath,
+): DropFirstScenePathElementResultType {
+  if (path.scene.sceneElementPaths.length === 0) {
```
Comment (Rheeseyb) [inline editor/src/core/shared/template-path.ts:934]: Should this use `isScenePathEmpty`?

File: editor/src/components/custom-code/code-file.spec.ts [hunk 6]
```diff
@@ -407,3 +442,179 @@ describe('Creating require function', () => {
...
+      ),
+    )
+    const expectedResult = normalisePathSuccess(
+      staticInstancePathFromString(':card-outer-div/card-inner-div'),
+      '/src/card.js',
+      getTextFileByPath(projectContents, '/src/card.js'),
+    )
+    expect(actualResult).toEqual(expectedResult)
+  })
+  it('returns the same path because there are no hops to take', () => {
```
Comment (Rheeseyb) [inline editor/src/components/custom-code/code-file.spec.ts:576]: Please can you add a test case for covering no hops with an instance path that also has a scene path, so that we can ensure that's not a weird potential edge case.

**Review by Rheeseyb — APPROVED** [review]

File: editor/src/components/custom-code/code-file.ts (hunk 2 above)
Comment (seanparsons) [reply editor/src/components/custom-code/code-file.ts:372]: Yes.

File: editor/src/components/custom-code/code-file.ts (hunk 3 above)
Comment (seanparsons) [reply editor/src/components/custom-code/code-file.ts:432]: Shouldn't do, as `animated` is the value in `baseVariable`, which would be the reference to find in the imports.

File: editor/src/core/shared/template-path.ts (hunk 5 above)
Comment (seanparsons) [reply editor/src/core/shared/template-path.ts:934]: It probably should actually, it did, then this function changed a couple of times and now it probably can again.

File: editor/src/components/custom-code/code-file.ts (hunk 1 above)
Comment (seanparsons) [reply editor/src/components/custom-code/code-file.ts:321]: Could probably do with specifying the file that it's halted at so that an error could be constructed from it.

Comment (seanparsons) [reply editor/src/components/custom-code/code-file.ts:321]: Fixed in 9bf80891.

File: editor/src/core/shared/template-path.ts (hunk 5 above)
Comment (seanparsons) [reply editor/src/core/shared/template-path.ts]: Fixed in 9bf80891.

File: editor/src/components/custom-code/code-file.spec.ts (hunk 6 above)
Comment (seanparsons) [reply editor/src/components/custom-code/code-file.spec.ts:600]: Fixed in 9bf80891.
